### PR TITLE
feat: [CDS-52615]: AWS SAM Publish Changes

### PR DESCRIPTION
**Pull request by alluvamsi-harness**
## Describe your changes
AWS SAM Publish Step Manager and Delegate Changes
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)

### CODE REVIEW

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+@OwnedBy(HarnessTeam.CDP)
+@Value
+@Builder
+@TypeAlias("awsSamPackageDataOutcom")
+@JsonTypeName("awsSamPackageDataOutcom")
+@RecasterAlias("io.harness.cdng.aws.sam.beans.AwsSamPackageDataOutcom")
+public class AwsSamBuildPackageDataOutcome implements Outcome, ExecutionSweepingOutput {
+  String templateFileContent;
+  String configFileContent;
+}
```
Comment (PragyeshMishra01) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java:30]: does it need to be Outcome also?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import lombok.Builder;
+import lombok.Value;
+import org.springframework.data.annotation.TypeAlias;
+
+@OwnedBy(HarnessTeam.CDP)
+@Value
+@Builder
+@TypeAlias("awsSamPackageDataOutcom")
+@JsonTypeName("awsSamPackageDataOutcom")
+@RecasterAlias("io.harness.cdng.aws.sam.beans.AwsSamPackageDataOutcom")
```
Comment (PragyeshMishra01) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java:26]: NIT: spell check

File: 930-delegate-tasks/src/main/java/io/harness/delegate/aws/sam/AwsSamPublishCommandTaskHandler.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+              + "AwsSamPublishRequest"));
+    }
+    AwsSamPublishRequest awsSamPublishRequest = (AwsSamPublishRequest) awsSamCommandRequest;
+    if (!(awsSamPublishRequest.getAwsSamInfraConfig() instanceof AwsSamInfraConfig)) {
+      throw new InvalidArgumentsException(Pair.of("AwsSamInfraConfig", "Must be instance of AwsSamInfraConfig"));
+    }
+
+    if (isEmpty(awsSamPublishRequest.getTemplateFileContent())) {
+      throw new InvalidArgumentsException("Invalid AwsSamPublishRequest");
+    }
```
Comment (PragyeshMishra01) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/aws/sam/AwsSamPublishCommandTaskHandler.java:69]: can we add better message?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/AwsCliDelegateTaskHelper.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      awsCommandFailure(executionLogCallback, command, response);
+    }
+
+    AwsCliStsAssumeRoleCommandOutputSchema responseOutput =
+        yamlUtils.read(response.getOutput(), AwsCliStsAssumeRoleCommandOutputSchema.class);
+    AwsCliCredentialConfig awsSamCredentialConfig = AwsCliCredentialConfig.builder()
+                                                        .provider("aws")
+                                                        .accessKey(responseOutput.getCredentials().getAccessKeyId())
+                                                        .secretKey(responseOutput.getCredentials().getSecretAccessKey())
+                                                        .build();
```
Comment (PragyeshMishra01) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/AwsCliDelegateTaskHelper.java:118]: make sure that we are not logging these aws keys anywhere as these are secrets

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/AwsSamPublishConfig.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+package io.harness.delegate.task.aws.sam;
+
+import lombok.Builder;
+import lombok.Value;
+
+@Value
+@Builder
+public class AwsSamPublishConfig {
+  String publishCommandOptions;
+}
```
Comment (PragyeshMishra01) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/AwsSamPublishConfig.java:17]: shall we handle expressions here?

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/request/AwsSamCommandRequest.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    List<ExecutionCapability> capabilities =
+        new ArrayList<>(EncryptedDataDetailsCapabilityHelper.fetchExecutionCapabilitiesForEncryptedDataDetails(
+            cloudProviderEncryptionDetails, maskingEvaluator));
+
+    AwsConnectorDTO awsConnectorDTO = awsSamInfraConfig.getAwsConnectorDTO();
+    capabilities.addAll(AwsCapabilityHelper.fetchRequiredExecutionCapabilities(awsConnectorDTO, maskingEvaluator));
+    CrossAccountAccessDTO crossAccountAccess = awsConnectorDTO.getCredential().getCrossAccountAccess();
+    if (crossAccountAccess != null && crossAccountAccess.getCrossAccountRoleArn() != null) {
+      capabilities.add(AwsCliInstallationCapability.builder().criteria("AWS Cli Installed").build());
```
Comment (PragyeshMishra01) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/request/AwsSamCommandRequest.java:52]: why is this check required? 
dont we need aws cli always?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java (hunk 2 above)
Comment (alluvamsi-harness) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java]: Done

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/request/AwsSamCommandRequest.java (hunk 6 above)
Comment (alluvamsi-harness) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/request/AwsSamCommandRequest.java:52]: removed the check

File: 930-delegate-tasks/src/main/java/io/harness/delegate/aws/sam/AwsSamPublishCommandTaskHandler.java (hunk 3 above)
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/aws/sam/AwsSamPublishCommandTaskHandler.java:69]: Done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java (hunk 1 above)
Comment (alluvamsi-harness) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/beans/AwsSamBuildPackageDataOutcome.java]: Removed Outcome

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/AwsCliDelegateTaskHelper.java (hunk 4 above)
Comment (alluvamsi-harness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/AwsCliDelegateTaskHelper.java:118]: Checked not logging them

**Review by rohitk-h — APPROVED** [review]
> Reviewed 43 of 48 files at r1, 6 of 23 files at r2, 41 of 57 files at r3, 21 of 21 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)* status: 71 file(s) to review. 4. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/AwsCliDelegateTaskHelper.java` line 48 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NOROR7m4t2Tveu-KB8g:-NOROR7m4t2Tveu-KB8h:bpxpoj6) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/930-delegate-tasks/src/main/java/io/harness/delegate/task/aws/AwsCliDelegateTaskHelper.java#L48)):*
> ```Java
>       AwsCliCredentialConfig awsCliCredentialConfig, long timeoutInMillis, Map<String, String> envVariables,
>       String workingDirectory, AwsCliConfig awsCliConfig) throws InterruptedException, IOException, TimeoutException {
>     configureAwsCredentialCommand(awsCliClient, executionLogCallback, "aws_access_key_id",
> ```

Let's make these strings as constants and prevent sonar issues :-) 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/awssam/AwsSamStepCommonHelper.java` line 47 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NORMhpNCMJYbBls89TE:-NORMhpNCMJYbBls89TF:b-z9tb63) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/125-cd-nextgen/src/main/java/io/harness/cdng/awssam/AwsSamStepCommonHelper.java#L47)):*
> ```Java
>                             .build();
>     String taskName = taskType.getDisplayName() + " : " + awsSamCommandRequest.getCommandName();
>     GoogleFunctionsSpecParameters googleFunctionsSpecParameters =
> ```

Copy paste error
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/awssam/publish/AwsSamPublishBaseStepInfo.java` line 41 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NORMslC4b9Vg-7h7psp:-NORMslC4b9Vg-7h7psq:b-g85bvh) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/125-cd-nextgen/src/main/java/io/harness/cdng/awssam/publish/AwsSamPublishBaseStepInfo.java#L41)):*
> ```Java
> 
>   @ApiModelProperty(dataType = SwaggerConstants.STRING_CLASSPATH) ParameterField<String> publishCommandOptions;
>   @JsonIgnore String awsSamPackageFnq;
> ```

Should this be Fqn instead? 
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/awssam/publish/AwsSamPublishStepInfo.java` line 44 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NORN_NOE6l3yRPGYutE:-NORN_NOE6l3yRPGYutF:b-ia41wb) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/125-cd-nextgen/src/main/java/io/harness/cdng/awssam/publish/AwsSamPublishStepInfo.java#L44)):*
> ```Java
>   ParameterField<List<TaskSelectorYaml>> delegateSelectors;
> 
>   @JsonIgnore String awsSamPackageFnq;
> ```

same comment as above
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/awssam/request/AwsSamPublishRequest.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NORQJ031OZB7rR8UZmU:-NORQJ031OZB7rR8UZmV:bycsbc4) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/awssam/request/AwsSamPublishRequest.java#L33)):*
> ```Java
>   String commandName;
>   CommandUnitsProgress commandUnitsProgress;
>   @NonFinal @Expression(ALLOW_SECRETS) AwsSamManifestConfig serverlessManifestConfig;
> ```

typo: update variable name to samManifestConfig
___
*[`960-api-services/src/main/java/io/harness/awssam/command/BUILD.bazel` line 16 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NOROrwrB3E9G4sZqBp6:-NOROrwrB3E9G4sZqBp7:bd454uf) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/960-api-services/src/main/java/io/harness/awssam/command/BUILD.bazel#L16)):*
> ```Starlark
> )
> 
> run_analysis_per_module()
> ```

if this is empty, can this be removed?

**Review by rohitk-h — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)* status: 71 file(s) to review. 4. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by alluvamsi-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)* status: 71 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/awssam/AwsSamStepCommonHelper.java` line 47 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NORMhpNCMJYbBls89TE:-NOmghY90T_53lXiKDEi:bo7k7gu) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/125-cd-nextgen/src/main/java/io/harness/cdng/awssam/AwsSamStepCommonHelper.java#L47)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

Copy paste error
</blockquote></details>

this is changed in later revisions
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/awssam/request/AwsSamPublishRequest.java` line 33 at r1](https://reviewable.io/reviews/harness/harness-core/44021#-NORQJ031OZB7rR8UZmU:-NOmrZiG-eB7meeeTjtJ:buvyehi) ([raw file](https://github.com/harness/harness-core/blob/8fe946fd6440ef8821a8884ce01f0f6679439b02/950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/awssam/request/AwsSamPublishRequest.java#L33)):*
<details><summary><i>Previously, rohitk-h (Rohit Karelia) wrote…</i></summary><blockquote>

typo: update variable name to samManifestConfig
</blockquote></details>

this was in older commit, changed it later commit

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/AwsSamPublishConfig.java (hunk 5 above)
Comment (sainath-b) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/AwsSamPublishConfig.java:22]: @alluvamsi-harness Please handle this

Comment (alluvamsi-harness) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/aws/sam/AwsSamPublishConfig.java:22]: Done

**Review by PragyeshMishra01 — APPROVED** [review]

**Review by sainath-b — APPROVED** [review]

**Review by rohitk-h — DISMISSED** [review]
> Reviewed 9 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)* status: 70 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by rohitk-h — APPROVED** [review]
> Reviewed 18 of 18 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)* status: 70 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by rohitk-h — APPROVED** [review]
> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)* status: 70 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by rohitk-h — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44021)* status: 70 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Comment by alluvamsi-harness** [comment]
trigger messagecheck

**Pull request closed without merging**